### ISSUE TITLE
Concept `SafeNumeric` is unnecessary

### ISSUE THREAD

**Issue by akrzemi1** [boostorg/safe_numerics#93]
This library defines concept `SafeNumeric` but it never uses it for anything. In generic libraries concepts are used as an interface between templates that require something of their types, and concrete types that want to adhere to these requirements. This is the case for concept `Numeric`: template `safe<>` requires it as a constraint, and types like `int` or `safe<int>` model it. In case of `SafeNumeric` some types model it, but no template requires it.

Also, intuitively, this library is for providing `safe<int>` which is *just like `int`*. This "just like `int`" is expressed with concept `Numeric`. One cannot apply an analogous reasoning to `SafeNumeric` and say, my function works only for things that are "just like `safe<int>`".

**Comment by robertramey** [comment]
>This library defines concept SafeNumeric but it never uses it for anything
If nothing else, it's used for documenting behavior of the library types.  
But it's more than that.  The type trait is_safe<T> is used through out the library.  In many/most cases it's used as an argument to enable_if to select appropriate definitions. But I know that at least in some cases, it is used to catch user errors in function calls and/or in user implementations of promotion and/or exception policies.

Of course I see that as I write this, is_safe<T> is not documented.  I will add this to the SafeNumeric concepts

You're correct of course that in the library, we don't really illustrate the usage fo SafeNumeric concept.  This is not a huge surprise since it's a "top level" type not used (very much) within the library.  But users that define their own types and/or functions which require safe parameters could/should/would use is_safe<T>::value.  Consider the example "Programming by Contract is Too Slow".  In this case test2 takes safe types as parameters:

auto test2(unsigned int hours, unsigned int minutes){
    // solution: use safe numerics
    // safe types can be implicitly constructed base types
    // construction guarentees corectness
    // return value is known to fit in unsigned int
    return convert(hours, minutes);
}

If one wanted catch user errors of passing an unsafe type, he could do the following instead.

auto test2(unsigned int hours, unsigned int minutes){
    // solution: use safe numerics
    static_assert(is_safe<hours>::value);
    static_assert(is_safe<minutes>::value);
    // safe types can be implicitly constructed base types
    // construction guarentees corectness
    // return value is known to fit in unsigned int
    return convert(hours, minutes);
}

Granted, not the greatest example as the library includes implicit conversion to safe types when appropriate.  But its the best example I can come up with on short notice.

> Also, intuitively, this library is for providing safe<int> which is just like int
I think it would be more accurate and helpful to rephrase this as:

"intuitively, this library is for providing safe<int> which is an enhance version of int".  Works like an int - but can never return an erroneous result.

**Comment by robertramey** [comment]
Damn - I see that is_safe is indeed documented in the SafeNumeric<T> concept.  So some of the above commentary is unnecessary.

**Comment by robertramey** [comment]
closing this - feel free to re-open if necessary.
